// @roochnetwork/rooch-sdk@0.2.2/dist/cjs/address/types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  MultiChainID: () => MultiChainID
});
module.exports = __toCommonJS(types_exports);
var MultiChainID = /* @__PURE__ */ ((MultiChainID2) => {
  MultiChainID2[MultiChainID2["Bitcoin"] = 0] = "Bitcoin";
  MultiChainID2[MultiChainID2["Ether"] = 60] = "Ether";
  MultiChainID2[MultiChainID2["Sui"] = 784] = "Sui";
  MultiChainID2[MultiChainID2["Nostr"] = 1237] = "Nostr";
  MultiChainID2[MultiChainID2["Rooch"] = 20230101] = "Rooch";
  return MultiChainID2;
})(MultiChainID || {});
