// @roochnetwork/rooch-sdk@0.2.2/dist/cjs/crypto/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var crypto_exports = {};
__export(crypto_exports, {
  Authenticator: () => import_authenticator.Authenticator,
  BitcoinSignMessage: () => import_authenticator.BitcoinSignMessage,
  BuiltinAuthValidator: () => import_authenticator.BuiltinAuthValidator,
  Keypair: () => import_keypair.Keypair,
  LEGACY_PRIVATE_KEY_SIZE: () => import_keypair.LEGACY_PRIVATE_KEY_SIZE,
  PRIVATE_KEY_SIZE: () => import_keypair.PRIVATE_KEY_SIZE,
  PublicKey: () => import_publickey.PublicKey,
  ROOCH_SECRET_KEY_PREFIX: () => import_keypair.ROOCH_SECRET_KEY_PREFIX,
  Signer: () => import_signer.Signer,
  decodeRoochSercetKey: () => import_keypair.decodeRoochSercetKey,
  encodeRoochSercetKey: () => import_keypair.encodeRoochSercetKey
});
module.exports = __toCommonJS(crypto_exports);
__reExport(crypto_exports, require("./mnemonics.js"), module.exports);
__reExport(crypto_exports, require("./signatureScheme.js"), module.exports);
var import_publickey = require("./publickey.js");
var import_keypair = require("./keypair.js");
var import_signer = require("./signer.js");
var import_authenticator = require("./authenticator.js");
