// @roochnetwork/rooch-sdk@0.2.2/dist/cjs/client/error.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var error_exports = {};
__export(error_exports, {
  ErrorValidateAccountDoesNotExist: () => ErrorValidateAccountDoesNotExist,
  ErrorValidateBadChainId: () => ErrorValidateBadChainId,
  ErrorValidateCantPayGasDeposit: () => ErrorValidateCantPayGasDeposit,
  ErrorValidateFunctionCallBeyondSessionScope: () => ErrorValidateFunctionCallBeyondSessionScope,
  ErrorValidateInvalidAccountAuthKey: () => ErrorValidateInvalidAccountAuthKey,
  ErrorValidateInvalidAuthenticator: () => ErrorValidateInvalidAuthenticator,
  ErrorValidateMaxGasAmountExceeded: () => ErrorValidateMaxGasAmountExceeded,
  ErrorValidateNotInstalledAuthValidator: () => ErrorValidateNotInstalledAuthValidator,
  ErrorValidateSequenceNuberTooOld: () => ErrorValidateSequenceNuberTooOld,
  ErrorValidateSequenceNumberTooBig: () => ErrorValidateSequenceNumberTooBig,
  ErrorValidateSequenceNumberTooNew: () => ErrorValidateSequenceNumberTooNew,
  ErrorValidateSessionIsExpired: () => ErrorValidateSessionIsExpired,
  ErrorValidateTransactionExpired: () => ErrorValidateTransactionExpired,
  JsonRpcError: () => JsonRpcError,
  RoochHTTPStatusError: () => RoochHTTPStatusError,
  RoochHTTPTransportError: () => RoochHTTPTransportError
});
module.exports = __toCommonJS(error_exports);
const ErrorValidateSequenceNuberTooOld = 1001;
const ErrorValidateSequenceNumberTooNew = 1002;
const ErrorValidateAccountDoesNotExist = 1003;
const ErrorValidateCantPayGasDeposit = 1004;
const ErrorValidateTransactionExpired = 1005;
const ErrorValidateBadChainId = 1006;
const ErrorValidateSequenceNumberTooBig = 1007;
const ErrorValidateMaxGasAmountExceeded = 1008;
const ErrorValidateInvalidAccountAuthKey = 1009;
const ErrorValidateInvalidAuthenticator = 1010;
const ErrorValidateNotInstalledAuthValidator = 1011;
const ErrorValidateSessionIsExpired = 1012;
const ErrorValidateFunctionCallBeyondSessionScope = 1013;
const CODE_TO_ERROR_TYPE = {
  1001: "SequenceNuberTooOld",
  1002: "SequenceNuberTooNew",
  1003: "AccountDoesNotExist",
  1004: "CantPayGasDeposit",
  1005: "TransactionExpired",
  1006: "BadChainId",
  1007: "SequenceNumberTooBig",
  1008: "MaxGasAmountExceeded",
  1009: "InvalidAccountAuthKey",
  1010: "InvalidAuthenticator",
  1011: "NotInstalledAuthValidator",
  1012: "SessionIsExpired",
  1013: "CallFunctionBeyondSessionScop"
};
class RoochHTTPTransportError extends Error {
}
class JsonRpcError extends RoochHTTPTransportError {
  constructor(message, code) {
    super(message);
    const parse = this.parseSubStatus();
    this.code = parse || code;
    this.type = CODE_TO_ERROR_TYPE[this.code] ?? "ServerError";
  }
  // Parse rooch RPC error sub status from `status ABORTED of type Execution with sub status 66537`
  parse() {
  }
  parseSubStatus() {
    const regex = /sub status (\d+)/;
    const match = this.message.match(regex);
    const code = match ? parseInt(match[1]) : null;
    return code ? code & 65535 : null;
  }
}
class RoochHTTPStatusError extends RoochHTTPTransportError {
  constructor(message, status, statusText) {
    super(message);
    this.status = status;
    this.statusText = statusText;
  }
}
